src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ExpressionTextCache.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    /// <summary>
+    /// This class holds the cache for the expression text that is computed by ExpressionHelper.
+    /// </summary>
+    public class ExpressionTextCache
+    {
+        public ConcurrentDictionary<LambdaExpression, string> Entries { get; } = 
+            new ConcurrentDictionary<LambdaExpression, string>(LambdaExpressionComparer.Instance);
+
+        private class LambdaExpressionComparer : IEqualityComparer<LambdaExpression>
+        {
+            public static readonly LambdaExpressionComparer Instance = new LambdaExpressionComparer();
+
+            public bool Equals(LambdaExpression lambdaExpression1, LambdaExpression lambdaExpression2)
+            {
+                if (ReferenceEquals(lambdaExpression1,lambdaExpression2))
+                {
+                    return true;
+                }
+                // We will cache only pure member access expressions. Hence we compare two expressions 
+                // to be equal only if they are identical member access expressions.
+                var expression1 = lambdaExpression1.Body;
+                var expression2 = lambdaExpression2.Body;
+                var isEquivalent = false;
+
+                while (true)
+                {
+                    if (expression1 == null || expression2 == null)
+                    {
+                        break;
+                    }
+
+                    if (expression1.NodeType != expression2.NodeType)
+                    {
+                        break;
+                    }
+
+                    if (expression1.NodeType == ExpressionType.MemberAccess)
+                    {
+                        var memberExpression1 = (MemberExpression)expression1;
+                        var memberName1 = memberExpression1.Member.Name;
+                        expression1 = memberExpression1.Expression;
+
+                        var memberExpression2 = (MemberExpression)expression2;
+                        var memberName2 = memberExpression2.Member.Name;
+                        expression2 = memberExpression2.Expression;
+
+                        // If identifier contains "__", it is "reserved for use by the implementation" and likely compiler-
+                        // or Razor-generated e.g. the name of a field in a delegate's generated class.
+                        if (memberName1.Contains("__") && memberName2.Contains("__"))
+                        {
+                            isEquivalent = true;
+                            break;
```

Comment:
This code would be much more understandable just returning a value instead of relying on `break` for control flow. `isEquivalent` is never set to true without a break, so you don't even need it.